stan/math/prim/functor/hcubature.hpp

```diff
@@ -0,0 +1,521 @@
+                              1.0479001032225018383987632254189e-01,
+                              1.4065325971552591874518959051021e-01,
+                              1.6900472663926790282658342659795e-01,
+                              1.9035057806478540991325640242055e-01,
+                              2.0443294007529889241416199923466e-01,
+                              2.0948214108472782801299917489173e-01};
+
+static const double gwd7[4] = {1.2948496616886969327061143267787e-01,
+                               2.797053914892766679014677714229e-01,
+                               3.8183005050511894495036977548818e-01,
+                               4.1795918367346938775510204081658e-01};
+
+struct one_d {
+  double result;
+  double err;
+  int kdivide = 0;
+};
+
+struct GenzMalik {
+  std::vector<double*> p[4];
+  double w[5];
+  double wd[4];
+};
+
+template <typename T_c, typename T_n, typename T_p, typename T_x>
+void combination(T_c& c, const T_n& n, const T_p& p, const T_x& x) {
+  size_t i, r, k = 0;
+  for (i = 0; i < p - 1; i++) {
+    c[i] = (i != 0) ? c[i - 1] : 0;
+    do {
+      c[i]++;
+      r = choose(n - c[i], p - (i + 1));
+      k = k + r;
+    } while (k < x);
+    k = k - r;
+  }
+  if (p > 1)
+    c[p - 1] = c[p - 2] + x - k;
+  else
+    c[0] = x;
+}
+
+template <typename T_k, typename T_lambda, typename T_n, typename T_p>
+void combos(const T_k& k, const T_lambda& lambda, const T_n& n, T_p& p) {
+  T_k* c = reinterpret_cast<T_k*>(malloc(k * sizeof(T_k)));
+  for (size_t i = 1; i != choose(n, k) + 1; i++) {
+    T_lambda* temp = reinterpret_cast<T_lambda*>(calloc(n, sizeof(T_lambda)));
+    combination(c, n, k, i);
+    for (size_t j = 0; j != k; j++)
+      temp[c[j] - 1] = lambda;
+    p.push_back(temp);
+  }
+  free(c);
+}
+
+template <typename T_index, typename T_k, typename T_lambda, typename T_n,
+          typename T_c, typename T_temp>
+void increment(T_index& index, const T_k& k, const T_lambda& lambda,
+               const T_n& n, const T_c& c, T_temp& temp) {
+  // temp size n, all elements initially zero
+  if (index.size() == 0) {
+    index.push_back(false);
+    for (size_t j = 0; j != k; j++)
+      temp[c[j] - 1] = lambda;
+    return;
+  }
+  T_k first_zero = 0;
+  while ((first_zero < index.size()) && index[first_zero])
+    first_zero++;
+  if (first_zero == index.size()) {
+    index.flip();
+    for (size_t j = 0; j != index.size(); j++)
+      temp[c[j] - 1] *= -1;
+    index.push_back(true);
+    temp[c[index.size() - 1] - 1] = -lambda;
+  } else {
+    for (size_t i = 0; i != first_zero + 1; i++) {
+      //           index[i] = !index[i];
+      if (index[i])
+        index[i] = 0;
+      else
+        index[i] = 1;
+      temp[c[i] - 1] *= -1;
+    }
+  }
+}
+
+template <typename T_k, typename T_lambda, typename T_n, typename T_p>
+void signcombos(const T_k& k, const T_lambda& lambda, const T_n& n, T_p& p) {
+  T_k* c = reinterpret_cast<T_k*>(malloc(k * sizeof(T_k)));
```

Comment:
Is `T_k` an integral type? If so then it might be nicer to use `std::vector<T_k>`. Or `Eigen::Array<T_k, -1, 1>`. In general though we avoid raw mallocs and frees